src/home/test-homed-regression-31896.c

```diff
@@ -0,0 +1,36 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "bus-locator.h"
+#include "fd-util.h"
```

Comment:
`fd-util.h` is not necessary?

Now, the test code simply call bus methods. Is it necessary that the test code is written in C? Why not simply call methods by e.g. `busctl`?

Reply:
busctl cannot hold on to FDs, which the bug requires to be reproduced

Reply:
Removed `fd-util.h`. It was a leftover from previous iterations of the test

Reply:
FWIW, I think C is a bit overkill here too. I personally would have forked off python to make calls over gdbus to keep it in the land of scripting languages, but :shrug: maybe that's incompatible w/ your testing environments. Not necessarily the prettiest solution either, but then again neither is this

Reply:
> busctl cannot hold on to FDs, which the bug requires to be reproduced

Hm, but the fd passed by `RefHomeUnrestricted` is not used in the test code? I may be confused...

Reply:
I mean, how homed checks if the bus client 'holds' the passed fd?

Reply:
It doesn't need to be used, it just needs to be held open

The kernel notifies homed when the FD returned by the dbus method is closed. This happens when the test client exits (actually when it frees the reply message immediately before exiting)

Reply:
Oh, interesting. I did not know there exists such feature. Thank you. LGTM.

Reply:
Basically you can open a pipe (or FIFO). Poll or read the read end of the pipe. It'll report EOF if the write end of the pipe is closed. You can duplicate the write end of the pipe many times over, and send the duplicates over IPC. When all the clients close their FDs, the write end of the pipe is closed by the kernel, thus when homed polls the read end it's told that the pipe is EOF and so homed gets notified that way.

This is how logind inhibitors work too. Basically acts as an inter-process reference count. The benefit is that clients can crash/exit without breaking the refcounting